extensions/eclipse-che-theia-plugin-ext/src/plugin/che-sidecar-file-system.ts

```diff
@@ -0,0 +1,223 @@
+/**********************************************************************
+ * Copyright (c) 2018-2020 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+import { CheSideCarFileSystem, FileTypeMain, PLUGIN_RPC_CONTEXT } from '../common/che-protocol';
+import { Stats, lstat, readFile, readdir, stat } from 'fs';
+
+import { RPCProtocol } from '@theia/plugin-ext/lib/common/rpc-protocol';
+import { URI } from 'vscode-uri';
+import { promisify } from 'util';
+
+export namespace SideCarFileSystemProvider {
+  export interface StatAndLink {
+    // The stats of the file. If the file is a symbolic
+    // link, the stats will be of that target file and
+    // not the link itself.
+    // If the file is a symbolic link pointing to a non
+    // existing file, the stat will be of the link and
+    // the `dangling` flag will indicate this.
+    stat: Stats;
+
+    // Will be provided if the resource is a symbolic link
+    // on disk. Use the `dangling` flag to find out if it
+    // points to a resource that does not exist on disk.
+    symbolicLink?: { dangling: boolean };
+  }
+}
+
+export enum FileSystemProviderErrorCode {
```

Comment:
Any reason we're not using the stuff from the `@theia/filesystem/common` module here?

Reply:
I've removed redundant code and reuse those objects from `common` module.